File: Part 2/B. Symmetric Matrix.cpp
// Solver : Nahin the Coder
// Date : 10 - 08 - 2021

#include<bits/stdc++.h>
#define ll long long int
using namespace std;

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	
    ll test;
    cin>>test;
    while(test--)
    {
    	bool flag = false;
    	int n,m;
    	cin>>n>>m;
    	int a,b,c,d;
    	for(int i=1; i<=n; i++)
    	{
    		cin>>a>>b>>c>>d;
    		if(b == c)
    		{
    			flag = true;
			}
		}
		if(m%2 !=0 or flag == false)
		{
			cout<<"NO"<<endl;
		}
		else
		{
			cout<<"YES"<<endl;
		}
    }
    return 0;
}

File: Solved by Python/B. Symmetric Matrix.py
# Nahin the Coder
# Date : 10 - 08 - 2021


testcase = int(input())
while testcase > 0:
    n, m = map(int, input().split())
    i = int(1)
    flag = bool(0)
    while i <= n:
        a, b = map(int, input().split())
        c, d = map(int, input().split())
        if b == c:
            flag = 1
        i = i + 1
    if m % 2 != 0 or flag == 0:
        print("NO")
    else:
        print("YES")
    testcase = testcase - 1

